File: Problem36_ContainerWithMostWater.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int maxArea(int[] height) {
        if(height == null || height.length == 0)
            return 0;
        int maxArea = Integer.MIN_VALUE;
        int left = 0, right = height.length - 1;
        while(left < right) {
            maxArea = Math.max(maxArea, Math.min(height[left], height[right]) * (right - left));
            if(height[left] < height[right])
                left++;
            else
                right--;
        }
        return maxArea;
    }
}

File: Problem 36 Container With Most Water.py
# Did this code successfully run on Leetcode: Yes
# Time Complexity: O(n)
# Space Complexity: O(1)
# Logic: put 1 pointer at start & one at end, calculate total water it holds with lower height. update pointers
# according to the whichever height is lower. return max water

class Solution:
    def maxArea(self, height):
        i = 0
        j = len(height) - 1
        max_water = 0
        while i < j:
            if height[i] < height[j]:
                max_water = max(max_water, (height[i]*(j-i)))
                i += 1
            else:
                max_water = max(max_water, (height[j] * (j - i)))
                j -= 1
        return max_water
